src/v/cloud_storage/remote_partition.cc

```diff
@@ -608,6 +608,7 @@ remote_partition::aborted_transactions(offset_range offsets) {
 }
 
 ss::future<> remote_partition::stop() {
+    _state = partition_state::stopping;
     vlog(_ctxlog.debug, "remote partition stop {} segments", _segments.size());
 
     co_await _gate.close();
```

Comment:
I believe that the hook is set up with "auto unlink" so it's not needed. It should be enough to remove all elements from the `_segments`.

Reply:
Just to be clear, the suggestion here is to move the segments from `_segments` to a temporary list, clear the original `_segments` and then stop the temporary list elements? 

Because I think we can't remove elements from the map while iterating it, as the materialized segments offload API could still call in and change the map.

Reply:
> Just to be clear, the suggestion here is to move the segments from _segments to a temporary list, clear the original _segments and then stop the temporary list elements?

Right, exactly.

Reply:
The the changes to maybe_trim_segment can go away, because we will always be stopping the segment from either the materialized_segments queue, or from partition::stop, but not both.

Reply:
got it, I will create a follow up PR for this today

Reply:
https://github.com/redpanda-data/redpanda/issues/7863 created

Reply:
@jcsp clearing out `_segments` might cause a problem in future. Considering the following scenario:

1. `materialized_segments::trim_segments` is iterating on segments and marks one of them for offloading.
2. It yields the core (this does not happen right now as I do not see a `co_await` but if the method changes it could, unless we can guarantee this method will always remain synchronous)
3. `remote_partition::stop` clears out the `_segments` map
4. Now when `materialized_segments::trim_segments` progresses it will try to offload the segment, where we have a vassert https://github.com/redpanda-data/redpanda/blob/4f9c498e96658f428a413702232ba3cef253ce45/src/v/cloud_storage/remote_partition.cc#L863 which will cause a crash.

Right now I think it is fine because `materialized_segments::trim_segments` does not yield anywhere. We could prevent this by making offloading a no-op if the partition is stopping but it starts getting complicated.

Reply:
> It yields the core (this does not happen right now as I do not see a co_await but if the method changes it could, unless we can guarantee this method will always remain synchronous)

I think we can commit to not making this method asynchronous in future: there are many places where we write logic that assumes certain places are synchronous.  We can add a comment to the function that makes it explicit that there must be no scheduling points in the loop.